tc/core/execution_engine_api_rfc.h

```diff
@@ -0,0 +1,149 @@
+ *   backend/backed_tc_executor.h and declares all the required dependent
+ *   **concrete** types.
+ * For example:
+ *   CudaBackend is declared in core/cuda/cuda_tc_executor.h
+ *
+ * struct CudaBackend {
+ *   using ExecutorType = CudaTcExecutor;
+ *   using MappingOptionsType = CudaMappingOptions;
+ *   using ProtobufType = CudaCacheProto;
+ *   using CacheType = CudaCache;
+ *   using CacheEntryType = CudaCacheEntry;
+ *   using CacheKeyType = typename CacheEntryType::Key;
+ *   using CacheValueType = typename CacheEntryType::Values;
+ *   using RTCFunctionType = CudaRTCFunction;
+ * };
+ *
+ * Sketching usage resembles:
+ *   std::string someTc = "...";
+ *   ExecutionEngine<CudaBackend> engine;
+ *   engine.define(someTc)
+ *   auto handle = engine.compile(someTc, inputs, mappingOptions)
+ *   auto cpuKernelTimings = engine.run(handle, inputs, outputs, true);
+ *   // alternatively:
+ *   // auto kernelTiming = engine.uncheckedRun(handle, inputs, outputs, true);
+ */
+template <typename Backend>
+class ExecutionEngine {
+ public:
+  using BackendType = Backend;
+  using ExecutorType = typename BackendType::ExecutorType;
+  using MappingOptionsType = typename BackendType::MappingOptionsType;
+
+  ExecutionEngine() = default;
+
+  /// Parse TC definitions provided as string, store parsed trees internally.
+  void define(const std::string& language);
+
+  /// Store the provided parsed trees internally.
+  void define(const std::vector<lang::TreeRef>& treeRefs);
+
+  /// For a given TC kernel name, compute the shapes of the output tensors
+  /// provided the shapes of the input TC tensors.  The caller can use the
+  /// computed shapes to allocate memory for outputs.
+  std::vector<TensorInfo> inferOutputTensorInfo(
+      const std::string& name,
+      const std::vector<const DLConstTensor*>& inTensorPtrs);
+
+  /// JIT-compile given the TC kernel name, the shapes of the input tensors and
+  /// the compilation options.  Must be overridden by a specific
+  /// ExecutionEngine, which also interprets the options as it sees fit.
+  /// \returns opaque handle of a compiled kernel.
+  size_t compile(
+      const std::string& name,
+      const std::vector<const DLConstTensor*>& inputs,
+      const MappingOptionsType& options);
+
+  /// Run a compiled TC kernel given its handle, on the given input tensors and
+  /// fill in the outputs.  All tensors must be allocated and have appropriate
+  /// shapes (inputs same as for copmilation, outputs same as returned by
+  /// inferOutputTensorInfo).
+  /// \returns The pair of (cpu,kernel) runtimes. If profile is set, the
+  /// kernel is profiled otherwise its duration is set to Duration::max().
+  std::pair<Duration, Duration> run(
+      size_t handle,
+      const std::vector<const DLConstTensor*>& inputs,
+      const std::vector<const DLTensor*>& outputs,
+      bool profileKernel = false,
+      std::function<bool(const ExecutorType*)> pruningFunction =
```

Comment:
I never really liked this callback here.  It's only use is in the autotuner deciding whether to run something or not.  It does not really make sense in the context of a single-short `run`, the caller should just decide.  Same for the autotuner by the way.  The logic seems inverted, but the caller just don't have enough information to take the decision.  See my proposal in a separate comment below.